### PR TITLE
[SPARK-10073] [SQL] Python withColumn should replace the old column

### PR DESCRIPTION
**Pull request by davies**
DataFrame.withColumn in Python should be consistent with the Scala one (replacing the existing column  that has the same name).

cc @marmbrus

### CODE REVIEW

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -1223,10 +1226,8 @@ def withColumnRenamed(self, existing, new):
         >>> df.withColumnRenamed('age', 'age2').collect()
         [Row(age2=2, name=u'Alice'), Row(age2=5, name=u'Bob')]
         """
-        cols = [Column(_to_java_column(c)).alias(new)
-                if c == existing else c
-                for c in self.columns]
-        return self.select(*cols)
+        assert existing in self.columns, "%s is not an existing column" % existing
```
Comment (cloud-fan) [inline python/pyspark/sql/dataframe.py]: This is not consistent with scala version, see the [doc](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala#L1173), we should not report error here.

Actually I'm wondering why we need to do checking at python side(not only this one)? Can we just call the scala API and catch the java exception?

Comment (davies) [inline python/pyspark/sql/dataframe.py]: Good catch, will remove it. I'm really surprised by the behavior in Scala.

The reason we want to have some check on Python side is that the Java exception is not easy  to understand for Python programmer (nested inside a Py4j exception). The Python exception or messages does improve the experience for Python programmer, especially beginners.

Comment (marmbrus) [inline python/pyspark/sql/dataframe.py]: Forgive my lack of python skills... could have have a generic wrapper that we use whenever we call back into scala that catches and unwraps certain expressions nicely?

Comment (davies) [inline python/pyspark/sql/dataframe.py]: We already did this for some of them, for example, AnalysisException and IllegalArgumentException, these could help.

Comment (marmbrus) [inline python/pyspark/sql/dataframe.py]: Oh cool, ideally we will pretty much only throw those two unless there is a bug.  If there are cases where that is not true we should consider fixing on the scala side.

**Comment by marmbrus** [comment]
Thanks!  Merging to master and 1.5.

**Pull request closed without merging**